src/graphql.ts

```diff
@@ -0,0 +1,68 @@
+import { default as fetch, Headers, RequestInit } from 'node-fetch'
+import { getAccessToken, getSourcegraphUrl } from './config'
+
+export const graphQLContent = Symbol('graphQLContent')
+export interface GraphQLDocument {
+    [graphQLContent]: string
+}
+
+/**
+ * Use this template string tag for all GraphQL queries
+ */
+export const gql = (template: TemplateStringsArray, ...substitutions: any[]): GraphQLDocument => ({
+    [graphQLContent]: String.raw(template, ...substitutions.map(s => s[graphQLContent] || s)),
+})
+
+export async function queryGraphQL(
+    graphQLDocument: GraphQLDocument,
+    variables: { [name: string]: any }
+): Promise<SourcegraphGQL.IQuery> {
+    return requestGraphQL(graphQLDocument, variables) as Promise<SourcegraphGQL.IQuery>
+}
+
+export async function mutateGraphQL(
+    graphQLDocument: GraphQLDocument,
+    variables: { [name: string]: any }
+): Promise<SourcegraphGQL.IMutation> {
+    return requestGraphQL(graphQLDocument, variables) as Promise<SourcegraphGQL.IMutation>
+}
+
+async function requestGraphQL(
+    graphQLDocument: GraphQLDocument,
+    variables: { [name: string]: any }
+): Promise<SourcegraphGQL.IQuery | SourcegraphGQL.IMutation> {
+    const headers = new Headers()
+    headers.append('User-Agent', 'Sourcegraph for Visual Studio Code')
+
+    const accessToken = getAccessToken()
+    if (accessToken) {
+        headers.append('Authorization', `token ${accessToken}`)
+    }
+
+    const query = graphQLDocument[graphQLContent]
+    const nameMatch = query.match(/^\s*(?:query|mutation)\s+(\w+)/)
+    const graphqlUrl = getSourcegraphUrl() + '/.api/graphql' + (nameMatch ? '?' + nameMatch[1] : '')
```

Comment:
It would be good to tack on `utm_source=editor` to these URLs so we can know if requests are coming from editors or something else in the future.

Reply:
We already know that from the `User-Agent` header, which seems like the better way to do this for the API compared to a query param (which is suitable for links cause it's the only way)

Reply:
@felixfbecker What exactly is the User-Agent for Visual Studio Code and how is it distinguishable from other browsers?

Reply:
Got it, I missed that line, thanks!